sql/src/Monitor.sk

```diff
@@ -0,0 +1,123 @@
+module alias P = SQLParser;
+
+module SKDBMonitor;
+
+const kReplicationWrites: String = "skdb.replication.writes";
+const kReplicationReads: String = "skdb.replication.reads";
+const kInvalidQueries: String = "skdb.queries.invalid";
+const kSucceedQueries: String = "skdb.queries.succeed";
+const kFailedQueries: String = "skdb.queries.failed";
+const kExecuted: String = "skdb.executed";
+const kBytesRead: String = "skdb.bytes.read";
+const kBytesWritten: String = "skdb.bytes.written";
+const kDatabase: String = "skdb.database";
+const kUser: String = "skdb.user";
+const kStmtType: String = "skdb.stmt.type";
+
+fun logStatement(stmt: P.Stmt): void {
+  SKMonitor.setStringAttribute(kStmtType, stmt.getClassName())
+}
+
+fun addReplicationWrites(writes: Int): void {
+  SKMonitor.incrIntAttribute(kReplicationWrites, writes)
+}
+
+fun addReplicationReads(writes: Int): void {
+  SKMonitor.incrIntAttribute(kReplicationReads, writes)
+}
+
+fun setName(name: String): void {
+  SKMonitor.setName(name)
+}
+
+fun setDatabase(name: String): void {
+  SKMonitor.setStringAttribute(kDatabase, name)
+}
+
+fun setUser(name: String): void {
+  SKMonitor.setStringAttribute(kUser, name)
+}
+
+fun addBytesRead(bytes: Int): void {
+  SKMonitor.incrIntAttribute(kBytesRead, bytes)
+}
+
+fun addBytesWritten(bytes: Int): void {
+  SKMonitor.incrIntAttribute(kBytesWritten, bytes)
+}
+
+fun start(): void {
+  SKMonitor.startSpan()
+}
+
+fun end(): void {
+  SKMonitor.endSpan()
+}
+
+module end;
+
+module SKDB;
+
+fun skipExit<T>(res: Int): T {
+  SKDBMonitor.end();
+  SKStore.destroy_Obstack(None());
+  .skipExit(res)
+}
+
+fun byteSize(value: String): Int {
+  String.byteSize(value).toInt()
+}
+
+fun print_error(error: String): void {
+  SKDBMonitor.addBytesWritten(byteSize(error));
+  .print_error(error);
+}
+
+fun print_raw(str: String): void {
+  SKDBMonitor.addBytesWritten(byteSize(str).toInt());
+  .print_raw(str);
+}
+
+fun print_string<T: readonly Show>(value: T): void {
+  .print_raw(value.toString());
+  .print_newline()
+}
+
+fun read_to_end(): String {
+  text = .read_to_end();
+  SKDBMonitor.addBytesRead(byteSize(text));
+  text
+}
+
+fun read_line(): ?String {
+  .read_line().map(line -> {
+    SKDBMonitor.addBytesRead(byteSize(line));
+    line
+  })
+}
+
+module end;
+
+module SKCSV;
+
+fun read_line(): ?String {
+  SKDB.read_line()
+}
+
+fun skipExit<T>(res: Int): T {
+  SKDB.skipExit(res)
+}
+
+fun print_error(error: String): void {
+  SKDB.print_error(error)
+}
+
+fun print_raw(str: String): void {
+  SKDB.print_raw(str)
+}
+
+fun print_string<T: readonly Show>(value: T): void {
+  SKDB.print_string(value)
+}
```

Comment:
Did you intend to capture stuff here?

Reply:
We need to manage the skipExit for clean exit and readline to get the number of read bytes.
The rest is to be belt and braces.